File: src/endstone_endweave/codec/types/level_settings.py
"""LevelSettings compound type for StartGamePacket."""

from dataclasses import dataclass

from endstone_endweave.codec.reader import PacketReader
from endstone_endweave.codec.writer import PacketWriter

from .gameplay import (
    EXPERIMENTS,
    EXPERIMENTS_V860,
    GAME_RULES,
    Experiment,
    GameRule,
)
from .primitives import (
    BOOL,
    BYTE,
    FLOAT_LE,
    INT64_LE,
    INT_LE,
    SHORT_LE,
    STRING,
    UVAR_INT,
    VAR_INT,
    OptionalType,
    Type,
)

_OPTIONAL_BOOL = OptionalType(BOOL)


@dataclass
class LevelSettings:
    """BDS LevelSettings struct (Seed through DisablePlayerInteractions)."""

    seed: int
    spawn_biome_type: int
    spawn_biome_name: str
    spawn_dimension: int
    generator: int
    game_type: int
    is_hardcore: bool
    game_difficulty: int
    default_spawn_x: int
    default_spawn_y: int
    default_spawn_z: int
    achievements_disabled: bool
    editor_world_type: int
    created_in_editor: bool
    exported_from_editor: bool
    day_cycle_stop_time: int
    education_edition_offer: int
    education_features_enabled: bool
    education_product_id: str
    rain_level: float
    lightning_level: float
    has_confirmed_platform_locked_content: bool
    multiplayer_intended: bool
    lan_broadcasting_intended: bool
    xbox_live_broadcast_setting: int
    platform_broadcast_setting: int
    commands_enabled: bool
    texture_packs_required: bool
    game_rules: list[GameRule]
    experiments: list[Experiment]
    ever_toggled: bool
    has_bonus_chest: bool
    start_with_map: bool
    player_permissions: int
    server_chunk_tick_range: int
    has_locked_behavior_pack: bool
    has_locked_resource_pack: bool
    is_from_locked_template: bool
    use_msa_gamertags_only: bool
    created_from_template: bool
    template_with_locked_settings: bool
    only_spawn_v1_villagers: bool
    persona_disabled: bool
    custom_skins_disabled: bool
    emote_chat_muted: bool
    base_game_version: str
    limited_world_width: int
    limited_world_depth: int
    nether_type: bool
    edu_shared_uri_button_name: str
    edu_shared_uri_link_uri: str
    override_force_experimental: bool | None
    chat_restriction_level: int
    disable_player_interactions: bool


class _LevelSettingsType(Type[LevelSettings]):
    """Version-parameterized LevelSettings reader/writer.

    Args:
        y_type: Type used for DefaultSpawn Y (UVAR_INT or VAR_INT).
        experiments_type: Type used for Experiments list.
    """

    def __init__(
        self,
        y_type: Type[int],
        experiments_type: Type[list[Experiment]],
    ) -> None:
        self._y_type = y_type
        self._experiments_type = experiments_type

    def read(self, reader: PacketReader) -> LevelSettings:
        return LevelSettings(
            seed=INT64_LE.read(reader),
            spawn_biome_type=SHORT_LE.read(reader),
            spawn_biome_name=STRING.read(reader),
            spawn_dimension=VAR_INT.read(reader),
            generator=VAR_INT.read(reader),
            game_type=VAR_INT.read(reader),
            is_hardcore=BOOL.read(reader),
            game_difficulty=VAR_INT.read(reader),
            default_spawn_x=VAR_INT.read(reader),
            default_spawn_y=self._y_type.read(reader),
            default_spawn_z=VAR_INT.read(reader),
            achievements_disabled=BOOL.read(reader),
            editor_world_type=VAR_INT.read(reader),
            created_in_editor=BOOL.read(reader),
            exported_from_editor=BOOL.read(reader),
            day_cycle_stop_time=VAR_INT.read(reader),
            education_edition_offer=VAR_INT.read(reader),
            education_features_enabled=BOOL.read(reader),
            education_product_id=STRING.read(reader),
            rain_level=FLOAT_LE.read(reader),
            lightning_level=FLOAT_LE.read(reader),
            has_confirmed_platform_locked_content=BOOL.read(reader),
            multiplayer_intended=BOOL.read(reader),
            lan_broadcasting_intended=BOOL.read(reader),
            xbox_live_broadcast_setting=VAR_INT.read(reader),
            platform_broadcast_setting=VAR_INT.read(reader),
            commands_enabled=BOOL.read(reader),
            texture_packs_required=BOOL.read(reader),
            game_rules=GAME_RULES.read(reader),
            experiments=self._experiments_type.read(reader),
            ever_toggled=BOOL.read(reader),
            has_bonus_chest=BOOL.read(reader),
            start_with_map=BOOL.read(reader),
            player_permissions=VAR_INT.read(reader),
            server_chunk_tick_range=INT_LE.read(reader),
            has_locked_behavior_pack=BOOL.read(reader),
            has_locked_resource_pack=BOOL.read(reader),
            is_from_locked_template=BOOL.read(reader),
            use_msa_gamertags_only=BOOL.read(reader),
            created_from_template=BOOL.read(reader),
            template_with_locked_settings=BOOL.read(reader),
            only_spawn_v1_villagers=BOOL.read(reader),
            persona_disabled=BOOL.read(reader),
            custom_skins_disabled=BOOL.read(reader),
            emote_chat_muted=BOOL.read(reader),
            base_game_version=STRING.read(reader),
            limited_world_width=INT_LE.read(reader),
            limited_world_depth=INT_LE.read(reader),
            nether_type=BOOL.read(reader),
            edu_shared_uri_button_name=STRING.read(reader),
            edu_shared_uri_link_uri=STRING.read(reader),
            override_force_experimental=_OPTIONAL_BOOL.read(reader),
            chat_restriction_level=BYTE.read(reader),
            disable_player_interactions=BOOL.read(reader),
        )

    def write(self, writer: PacketWriter, value: LevelSettings) -> None:
        INT64_LE.write(writer, value.seed)
        SHORT_LE.write(writer, value.spawn_biome_type)
        STRING.write(writer, value.spawn_biome_name)
        VAR_INT.write(writer, value.spawn_dimension)
        VAR_INT.write(writer, value.generator)
        VAR_INT.write(writer, value.game_type)
        BOOL.write(writer, value.is_hardcore)
        VAR_INT.write(writer, value.game_difficulty)
        VAR_INT.write(writer, value.default_spawn_x)
        self._y_type.write(writer, value.default_spawn_y)
        VAR_INT.write(writer, value.default_spawn_z)
        BOOL.write(writer, value.achievements_disabled)
        VAR_INT.write(writer, value.editor_world_type)
        BOOL.write(writer, value.created_in_editor)
        BOOL.write(writer, value.exported_from_editor)
        VAR_INT.write(writer, value.day_cycle_stop_time)
        VAR_INT.write(writer, value.education_edition_offer)
        BOOL.write(writer, value.education_features_enabled)
        STRING.write(writer, value.education_product_id)
        FLOAT_LE.write(writer, value.rain_level)
        FLOAT_LE.write(writer, value.lightning_level)
        BOOL.write(writer, value.has_confirmed_platform_locked_content)
        BOOL.write(writer, value.multiplayer_intended)
        BOOL.write(writer, value.lan_broadcasting_intended)
        VAR_INT.write(writer, value.xbox_live_broadcast_setting)
        VAR_INT.write(writer, value.platform_broadcast_setting)
        BOOL.write(writer, value.commands_enabled)
        BOOL.write(writer, value.texture_packs_required)
        GAME_RULES.write(writer, value.game_rules)
        self._experiments_type.write(writer, value.experiments)
        BOOL.write(writer, value.ever_toggled)
        BOOL.write(writer, value.has_bonus_chest)
        BOOL.write(writer, value.start_with_map)
        VAR_INT.write(writer, value.player_permissions)
        INT_LE.write(writer, value.server_chunk_tick_range)
        BOOL.write(writer, value.has_locked_behavior_pack)
        BOOL.write(writer, value.has_locked_resource_pack)
        BOOL.write(writer, value.is_from_locked_template)
        BOOL.write(writer, value.use_msa_gamertags_only)
        BOOL.write(writer, value.created_from_template)
        BOOL.write(writer, value.template_with_locked_settings)
        BOOL.write(writer, value.only_spawn_v1_villagers)
        BOOL.write(writer, value.persona_disabled)
        BOOL.write(writer, value.custom_skins_disabled)
        BOOL.write(writer, value.emote_chat_muted)
        STRING.write(writer, value.base_game_version)
        INT_LE.write(writer, value.limited_world_width)
        INT_LE.write(writer, value.limited_world_depth)
        BOOL.write(writer, value.nether_type)
        STRING.write(writer, value.edu_shared_uri_button_name)
        STRING.write(writer, value.edu_shared_uri_link_uri)
        _OPTIONAL_BOOL.write(writer, value.override_force_experimental)
        BYTE.write(writer, value.chat_restriction_level)
        BOOL.write(writer, value.disable_player_interactions)


LEVEL_SETTINGS_V860 = _LevelSettingsType(UVAR_INT, EXPERIMENTS_V860)
LEVEL_SETTINGS_V924 = _LevelSettingsType(UVAR_INT, EXPERIMENTS)
LEVEL_SETTINGS_V944 = _LevelSettingsType(VAR_INT, EXPERIMENTS)

File: src/codec/types/level_settings.cpp
/// LevelSettings codec implementation.

#include "endweave/codec/types/level_settings.h"
#include "endweave/codec/types/primitives.h"

namespace endweave {

namespace {

/// Read spawn Y as either uvarint (v860/v924) or varint (v944).
template <bool UseVarint>
std::expected<LevelSettings, ReadError> read_level_settings(PacketReader& r,
    std::expected<std::vector<Experiment>, ReadError>(*read_exps)(PacketReader&))
{
    LevelSettings s;
    #define TRY(field, expr) { auto _v = (expr); if (!_v) return std::unexpected(_v.error()); s.field = *_v; }

    TRY(seed, r.read_int64_le())
    TRY(spawn_biome_type, r.read_short_le())
    TRY(spawn_biome_name, r.read_string())
    TRY(spawn_dimension, r.read_varint())
    TRY(generator, r.read_varint())
    TRY(game_type, r.read_varint())
    TRY(is_hardcore, r.read_bool())
    TRY(game_difficulty, r.read_varint())
    TRY(default_spawn_x, r.read_varint())
    if constexpr (UseVarint) {
        TRY(default_spawn_y, r.read_varint())
    } else {
        auto y = r.read_uvarint();
        if (!y) return std::unexpected(y.error());
        s.default_spawn_y = static_cast<std::int32_t>(*y);
    }
    TRY(default_spawn_z, r.read_varint())
    TRY(achievements_disabled, r.read_bool())
    TRY(editor_world_type, r.read_varint())
    TRY(created_in_editor, r.read_bool())
    TRY(exported_from_editor, r.read_bool())
    TRY(day_cycle_stop_time, r.read_varint())
    TRY(education_edition_offer, r.read_varint())
    TRY(education_features_enabled, r.read_bool())
    TRY(education_product_id, r.read_string())
    TRY(rain_level, r.read_float_le())
    TRY(lightning_level, r.read_float_le())
    TRY(has_confirmed_platform_locked_content, r.read_bool())
    TRY(multiplayer_intended, r.read_bool())
    TRY(lan_broadcasting_intended, r.read_bool())
    TRY(xbox_live_broadcast_setting, r.read_varint())
    TRY(platform_broadcast_setting, r.read_varint())
    TRY(commands_enabled, r.read_bool())
    TRY(texture_packs_required, r.read_bool())

    { auto v = r.read<game_rules>(); if (!v) return std::unexpected(v.error()); s.game_rules = std::move(*v); }
    { auto v = read_exps(r); if (!v) return std::unexpected(v.error()); s.exps = std::move(*v); }

    TRY(ever_toggled, r.read_bool())
    TRY(has_bonus_chest, r.read_bool())
    TRY(start_with_map, r.read_bool())
    TRY(player_permissions, r.read_varint())
    TRY(server_chunk_tick_range, r.read_int_le())
    TRY(has_locked_behavior_pack, r.read_bool())
    TRY(has_locked_resource_pack, r.read_bool())
    TRY(is_from_locked_template, r.read_bool())
    TRY(use_msa_gamertags_only, r.read_bool())
    TRY(created_from_template, r.read_bool())
    TRY(template_with_locked_settings, r.read_bool())
    TRY(only_spawn_v1_villagers, r.read_bool())
    TRY(persona_disabled, r.read_bool())
    TRY(custom_skins_disabled, r.read_bool())
    TRY(emote_chat_muted, r.read_bool())
    TRY(base_game_version, r.read_string())
    TRY(limited_world_width, r.read_int_le())
    TRY(limited_world_depth, r.read_int_le())
    TRY(nether_type, r.read_bool())
    TRY(edu_shared_uri_button_name, r.read_string())
    TRY(edu_shared_uri_link_uri, r.read_string())

    // Optional<bool>
    {
        auto has = r.read_bool();
        if (!has) return std::unexpected(has.error());
        if (*has) {
            auto v = r.read_bool();
            if (!v) return std::unexpected(v.error());
            s.override_force_experimental = *v;
        }
    }

    TRY(chat_restriction_level, r.read_byte())
    TRY(disable_player_interactions, r.read_bool())

    #undef TRY
    return s;
}

template <bool UseVarint>
void write_level_settings(PacketWriter& w, const LevelSettings& s,
    void(*write_exps)(PacketWriter&, const std::vector<Experiment>&))
{
    w.write_int64_le(s.seed);
    w.write_short_le(s.spawn_biome_type);
    w.write_string(s.spawn_biome_name);
    w.write_varint(s.spawn_dimension);
    w.write_varint(s.generator);
    w.write_varint(s.game_type);
    w.write_bool(s.is_hardcore);
    w.write_varint(s.game_difficulty);
    w.write_varint(s.default_spawn_x);
    if constexpr (UseVarint)
        w.write_varint(s.default_spawn_y);
    else
        w.write_uvarint(static_cast<std::uint32_t>(s.default_spawn_y));
    w.write_varint(s.default_spawn_z);
    w.write_bool(s.achievements_disabled);
    w.write_varint(s.editor_world_type);
    w.write_bool(s.created_in_editor);
    w.write_bool(s.exported_from_editor);
    w.write_varint(s.day_cycle_stop_time);
    w.write_varint(s.education_edition_offer);
    w.write_bool(s.education_features_enabled);
    w.write_string(s.education_product_id);
    w.write_float_le(s.rain_level);
    w.write_float_le(s.lightning_level);
    w.write_bool(s.has_confirmed_platform_locked_content);
    w.write_bool(s.multiplayer_intended);
    w.write_bool(s.lan_broadcasting_intended);
    w.write_varint(s.xbox_live_broadcast_setting);
    w.write_varint(s.platform_broadcast_setting);
    w.write_bool(s.commands_enabled);
    w.write_bool(s.texture_packs_required);
    w.write<game_rules>(s.game_rules);
    write_exps(w, s.exps);
    w.write_bool(s.ever_toggled);
    w.write_bool(s.has_bonus_chest);
    w.write_bool(s.start_with_map);
    w.write_varint(s.player_permissions);
    w.write_int_le(s.server_chunk_tick_range);
    w.write_bool(s.has_locked_behavior_pack);
    w.write_bool(s.has_locked_resource_pack);
    w.write_bool(s.is_from_locked_template);
    w.write_bool(s.use_msa_gamertags_only);
    w.write_bool(s.created_from_template);
    w.write_bool(s.template_with_locked_settings);
    w.write_bool(s.only_spawn_v1_villagers);
    w.write_bool(s.persona_disabled);
    w.write_bool(s.custom_skins_disabled);
    w.write_bool(s.emote_chat_muted);
    w.write_string(s.base_game_version);
    w.write_int_le(s.limited_world_width);
    w.write_int_le(s.limited_world_depth);
    w.write_bool(s.nether_type);
    w.write_string(s.edu_shared_uri_button_name);
    w.write_string(s.edu_shared_uri_link_uri);
    // Optional<bool>
    if (s.override_force_experimental.has_value()) {
        w.write_bool(true);
        w.write_bool(*s.override_force_experimental);
    } else {
        w.write_bool(false);
    }
    w.write_byte(s.chat_restriction_level);
    w.write_bool(s.disable_player_interactions);
}

std::expected<std::vector<Experiment>, ReadError> read_exps_v860(PacketReader& r) { return r.read<experiments_v860>(); }
std::expected<std::vector<Experiment>, ReadError> read_exps_v924(PacketReader& r) { return r.read<experiments>(); }
void write_exps_v860(PacketWriter& w, const std::vector<Experiment>& v) { w.write<experiments_v860>(v); }
void write_exps_v924(PacketWriter& w, const std::vector<Experiment>& v) { w.write<experiments>(v); }

} // namespace

template <> auto PacketReader::read<level_settings_v860>() -> std::expected<LevelSettings, ReadError> {
    return read_level_settings<false>(*this, read_exps_v860);
}
template <> auto PacketReader::read<level_settings_v924>() -> std::expected<LevelSettings, ReadError> {
    return read_level_settings<false>(*this, read_exps_v924);
}
template <> auto PacketReader::read<level_settings_v944>() -> std::expected<LevelSettings, ReadError> {
    return read_level_settings<true>(*this, read_exps_v924);
}
template <> void PacketWriter::write<level_settings_v860>(const LevelSettings& val) {
    write_level_settings<false>(*this, val, write_exps_v860);
}
template <> void PacketWriter::write<level_settings_v924>(const LevelSettings& val) {
    write_level_settings<false>(*this, val, write_exps_v924);
}
template <> void PacketWriter::write<level_settings_v944>(const LevelSettings& val) {
    write_level_settings<true>(*this, val, write_exps_v924);
}

} // namespace endweave
